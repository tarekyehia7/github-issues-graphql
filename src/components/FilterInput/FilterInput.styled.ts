import styled from "styled-components";
import { CloseIcon } from "../../icons/CloseIcon";
import { SearchIcon } from "../../icons/SearchIcon";

export const Input = styled.input`
	padding: 5px 26px;
	font-size: 14px;
	line-height: 20px;
	vertical-align: middle;
	background-repeat: no-repeat;
	background-position: right 8px center;
	border: 1px solid #d0d7de;
	border-radius: 6px;
	outline: none;
	width: 100%;
	background-color: #f6f8fa;
	color: #57606a;
	box-shadow: inset 0 1px 0 rgba(208, 215, 222, 0.2);
`;

export const FilterContainer = styled.div`
	position: relative;
	display: flex;
	flex-direction: row;
	align-items: center;
`;

export const SearchIconStyled = styled(SearchIcon)`
	position: absolute;
	display: block;
	color: #57606a;
	text-align: center;
	pointer-events: none;
	fill: #ffffff;
	stroke-width: 0.1rem;
	stroke: #57606a;
	margin-left: 0.3rem;
`;

export const CloseIconStyled = styled(CloseIcon)`
	width: 18px;
	height: 18px;
	padding: 1px;
	margin-right: 4px;
	fill: #ffffff;
	text-align: center;
	background-color: #6e7781;
	border-radius: 6px;
`;

export const ClearSearchContainer = styled.div`
	display: flex;
	flex-direction: row;
	align-items: center;
	cursor: pointer;
	font-weight: 600;
	color: #57606a;
	padding-top: 1rem;
	&:hover {
		color: #0969da;
		text-decoration: none;
		${CloseIconStyled} {
			background: #0969da;
		}
	}
`;
