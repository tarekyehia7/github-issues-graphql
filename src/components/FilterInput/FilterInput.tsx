import React from 'react';

import {
	FilterContainer,
	ClearSearchContainer,
	CloseIconStyled,
	Input,
	SearchIconStyled,
} from './FilterInput.styled';

type FilterInputProps = {
	value: string;
	showClearHistoryText: boolean;
	onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
	onKeyUp: (e: React.KeyboardEvent<HTMLInputElement>) => void;
	onClearSearchHistoryClick: () => void;
};

export const FilterInput = ({
	value,
	onChange,
	onKeyUp,
	showClearHistoryText,
	onClearSearchHistoryClick,
}: FilterInputProps) => {
	return (
		<>
			<FilterContainer>
				<Input value={value} onKeyUp={onKeyUp} onChange={onChange} />
				<SearchIconStyled />
			</FilterContainer>
			{showClearHistoryText && (
				<ClearSearchContainer onClick={onClearSearchHistoryClick}>
					<CloseIconStyled /> Clear current search query, filters, and sorts
				</ClearSearchContainer>
			)}
		</>
	);
};
