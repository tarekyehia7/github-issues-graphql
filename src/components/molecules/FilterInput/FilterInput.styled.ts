import styled from 'styled-components';
import { CloseIcon } from '../../../icons/CloseIcon';
import { SearchIcon } from '../../../icons/SearchIcon';

export const FilterContainer = styled.div`
	position: relative;
	display: flex;
	flex-direction: row;
	align-items: center;
`;

export const SearchIconStyled = styled(SearchIcon)`
	position: absolute;
	display: block;
	color: ${({ theme }) => theme.colors.darkGray};
	text-align: center;
	pointer-events: none;
	fill: ${({ theme }) => theme.colors.white};
	stroke-width: 0.1rem;
	stroke: ${({ theme }) => theme.colors.darkGray};
	margin-left: 0.3rem;
`;

export const CloseIconStyled = styled(CloseIcon)`
	width: 18px;
	height: 18px;
	padding: 1px;
	margin-right: 4px;
	fill: ${({ theme }) => theme.colors.white};
	text-align: center;
	background-color: #6e7781;
	border-radius: 6px;
`;
