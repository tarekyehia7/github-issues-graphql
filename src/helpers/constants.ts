export const constants = {
    repository: 'facebook',
    projectName: 'react',
	searchTerm: 'title,body',
    issuesPerPage: 10,
    githubUrl: 'https://github.com'
};