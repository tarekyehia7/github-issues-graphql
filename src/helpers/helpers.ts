import formatDistanceToNowStrict from 'date-fns/formatDistanceToNowStrict';

export const getPagesNumber = (totalIssuesCount: number, issuesPerPage: number): number => {
	const totalPages = Math.ceil(totalIssuesCount / issuesPerPage);
	return !totalPages ? 1 : totalPages;
};

export const formatDate = (date: string) => {
	try {
		const myDate = new Date(date);
		const myFormattedDate = formatDistanceToNowStrict(myDate, { addSuffix: true });
		return myFormattedDate;
	} catch (e) {
		console.error(e);
		return '';
	}
};
