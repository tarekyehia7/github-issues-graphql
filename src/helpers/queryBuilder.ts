import { StatusEnum } from "../pages/Issues/Issues";

interface Query {
    repo: string;
    input: string;
    is: string;
    in: string;
    status: StatusEnum;
}

export const buildQuery = (query: Query) => {
    return Object.entries(query)
        .reduce((str, [p, val]) => `${str}${p}:${val} `, '')
        .replace('status:', '')
        .replace('input:', '')
        .trim();
};