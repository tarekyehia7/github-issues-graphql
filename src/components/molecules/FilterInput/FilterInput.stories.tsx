import React, { useEffect, useState } from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import { FilterInput } from './FilterInput';

export default {
	title: 'molecules/FilterInput',
	component: FilterInput,
	argTypes: {
		onChange: { action: 'changed' },
		onKeyUp: { action: 'pressed' },
		onClearSearchHistoryClick: { action: 'clicked' },
	},
} as ComponentMeta<typeof FilterInput>;

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
export const Primary: ComponentStory<typeof FilterInput> = args => {
	const [localValue, setValue] = useState<string>(args.value);
	const [localShowClearHistorytext, setLocalShowClearHistoryText] = useState(
		args.showClearHistoryText,
	);

	useEffect(() => {
		setValue(args.value);
	}, [args.value]);

	useEffect(() => {
		setLocalShowClearHistoryText(args.showClearHistoryText);
	}, [args.showClearHistoryText]);

	const onChangeInput = (e: React.ChangeEvent<HTMLInputElement>) => {
		setValue(e.target.value);
	};
	const onKeyUpInput = (e: React.KeyboardEvent<HTMLInputElement>) => {
		if (e.key === 'Enter') {
			setLocalShowClearHistoryText(true);
		}
	};

	const onClearSearchHistoryClick = () => {
		setValue('');
		setLocalShowClearHistoryText(false);
	};

	return (
		<FilterInput
			value={localValue}
			showClearHistoryText={localShowClearHistorytext}
			onChange={onChangeInput}
			onKeyUp={onKeyUpInput}
			onClearSearchHistoryClick={onClearSearchHistoryClick}
		/>
	);
};

Primary.args = {
	value: '',
	showClearHistoryText: false,
};
