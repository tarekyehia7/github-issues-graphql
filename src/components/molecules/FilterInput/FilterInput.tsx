import React from 'react';

import { FilterContainer, CloseIconStyled, SearchIconStyled } from './FilterInput.styled';
import { Input } from '../../atoms/Input/Input';
import { Button, ButtonType } from '../../atoms/button/Button';

export type FilterInputProps = {
	value: string;
	showClearHistoryText: boolean;
	onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
	onKeyUp: (e: React.KeyboardEvent<HTMLInputElement>) => void;
	onClearSearchHistoryClick: () => void;
};

export const FilterInput = ({
	value,
	onChange,
	onKeyUp,
	showClearHistoryText,
	onClearSearchHistoryClick,
}: FilterInputProps) => {
	return (
		<>
			<FilterContainer>
				<Input
					value={value}
					placeholder="type here..."
					onKeyUp={onKeyUp}
					onChange={onChange}
				/>
				<SearchIconStyled />
			</FilterContainer>
			{showClearHistoryText && (
				<Button disabled={false} type={ButtonType.Text} onClick={onClearSearchHistoryClick}>
					<CloseIconStyled /> Clear current search query, filters, and sorts
				</Button>
			)}
		</>
	);
};
