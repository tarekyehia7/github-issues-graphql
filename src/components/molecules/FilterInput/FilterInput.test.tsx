import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';

import { FilterInput, FilterInputProps } from './FilterInput';
import { PageWithTheme } from '../../../helpers/testing/helpers';

const mockOnChange = jest.fn();
const mockOnKeyUp = jest.fn();
const mockOnClearSearchHistoryClick = jest.fn();

const props: FilterInputProps = {
	value: 'this is my input',
	showClearHistoryText: true,
	onChange: mockOnChange,
	onKeyUp: mockOnKeyUp,
	onClearSearchHistoryClick: mockOnClearSearchHistoryClick,
};

const renderPage = () => {
	const { container, getByText, getByTestId } = render(
		<PageWithTheme>
			<FilterInput {...props} />
		</PageWithTheme>,
	);

	return {
		container,
		getByText,
		getByTestId,
	};
};

describe('<FilterInput />', () => {
	it('Should matches snapshot', () => {
		const { container } = renderPage();
		expect(container).toMatchSnapshot();
	});

	it('Should trigger onChange', () => {
		const { getByTestId } = renderPage();

		fireEvent.change(getByTestId('Input'), { target: { value: '23' } });
		expect(mockOnChange.mock.calls.length).toBe(1);
	});

	it('Should trigger onKeyUp', () => {
		const { getByTestId } = renderPage();

		fireEvent.keyUp(getByTestId('Input'));
		expect(mockOnKeyUp.mock.calls.length).toBe(1);
	});

	it('Should trigger click event on clicking clear history', () => {
		const { getByText } = renderPage();

		fireEvent.click(getByText('Clear current search query, filters, and sorts'));
		expect(mockOnClearSearchHistoryClick.mock.calls.length).toBe(1);
	});
});
