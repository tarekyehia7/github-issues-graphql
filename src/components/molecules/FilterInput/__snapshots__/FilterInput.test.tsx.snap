// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterInput /> Should matches snapshot 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  position: absolute;
  display: block;
  color: #57606a;
  text-align: center;
  pointer-events: none;
  fill: #ffffff;
  stroke-width: 0.1rem;
  stroke: #57606a;
  margin-left: 0.3rem;
}

.c4 {
  width: 18px;
  height: 18px;
  padding: 1px;
  margin-right: 4px;
  fill: #ffffff;
  text-align: center;
  background-color: #6e7781;
  border-radius: 6px;
}

.c1 {
  padding: 5px 26px;
  font-size: 14px;
  line-height: 20px;
  vertical-align: middle;
  background-repeat: no-repeat;
  background-position: right 8px center;
  border: 1px solid #d0d7de;
  border-radius: 6px;
  outline: none;
  width: 100%;
  background-color: #f6f8fa;
  box-shadow: inset 0 1px 0 rgba(208,215,222,0.2);
}

.c3 {
  color: #0969da;
  min-width: 2rem;
  padding: 5px 10px;
  font-style: normal;
  line-height: 20px;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 1px solid transparent;
  border-radius: 6px;
  -webkit-transition: border-color 0.2s cubic-bezier(0.3,0,0.5,1);
  transition: border-color 0.2s cubic-bezier(0.3,0,0.5,1);
  background-color: transparent;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  font-weight: 600;
  color: #57606a;
  padding-top: 1rem;
  background: transparent;
  border: none;
}

.c3:disabled {
  color: #8c959f;
  cursor: default;
  border-color: transparent;
}

.c3:disabled:hover {
  border-color: transparent;
}

.c3:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  border-color: #d0d7de;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
}

.c3:hover {
  color: #0969da;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c3:hover > svg {
  background: #0969da;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      data-testid="Input"
      placeholder="type here..."
      value="this is my input"
    />
    <svg
      aria-hidden="true"
      class="c2"
      data-view-component="true"
      height="16"
      version="1.1"
      viewBox="0 0 16 16"
      width="16"
    >
      <path
        d="M11.5 7a4.499 4.499 0 11-8.998 0A4.499 4.499 0 0111.5 7zm-.82 4.74a6 6 0 111.06-1.06l3.04 3.04a.75.75 0 11-1.06 1.06l-3.04-3.04z"
      />
    </svg>
  </div>
  <button
    class="c3"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="c4"
      data-view-component="true"
      version="1.1"
      viewBox="0 0 16 16"
    >
      <path
        d="M3.72 3.72a.75.75 0 011.06 0L8 6.94l3.22-3.22a.75.75 0 111.06 1.06L9.06 8l3.22 3.22a.75.75 0 11-1.06 1.06L8 9.06l-3.22 3.22a.75.75 0 01-1.06-1.06L6.94 8 3.72 4.78a.75.75 0 010-1.06z"
      />
    </svg>
     Clear current search query, filters, and sorts
  </button>
</div>
`;
